File: CS/Lib/CMCommon/DA/CMCmdSettings.cs
/*******************************************************************************
 * 【共通部品】
 *
 * 作成者: 日進テクノロジー／田中 望
 * 改版履歴:
 * 2014.1.30, 新規作成
 ******************************************************************************/
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Text;
using System.Web;

using NEXS.ERP.CM.Common;

namespace NEXS.ERP.CM.DA
{
    //************************************************************************
    /// <summary>
    /// SqlCommand設定コレクションクラス
    /// </summary>
    //************************************************************************
    public class CMCmdSettings
    {
        #region プロパティ
        [Category("共通部品")]
        [Description("SqlCommand設定のコレクション")]
        public List<CMCmdSetting> CmdSettings { get; set; }
        #endregion

        #region インデクサ
        /// <summary>指定IndexのSqlCommand設定を返します。</summary>
        public CMCmdSetting this[int argIndex]
        {
            get
            {
                return CmdSettings[argIndex];
            }
        }

        /// <summary>指定のテーブル名のSqlCommand設定を返します。</summary>
        public CMCmdSetting this[string argName]
        {
            get
            {
                var result = from row in CmdSettings
                             where row.Name == argName
                             select row;
                return result.First();
            }
        }
        #endregion

        #region コンストラクタ
        //************************************************************************
        /// <summary>
        /// コンストラクタ
        /// </summary>
        //************************************************************************
        public CMCmdSettings()
        {
            CmdSettings = new List<CMCmdSetting>();
        }

        //************************************************************************
        /// <summary>
        /// コンストラクタ
        /// </summary>
        /// <param name="argFname">読み込むXMLファイル名(拡張子なし)</param>
        //************************************************************************
        public CMCmdSettings(string argFname) : this()
        {
            AddFomXml(argFname);
        }
        #endregion

        //************************************************************************
        /// <summary>
        /// SqlCommand設定をXMLファイルから追加する。
        /// </summary>
        /// <param name="argFnames">読み込むXMLファイル名(拡張子なし)</param>
        //************************************************************************
        public void AddFomXml(params string[] argFnames)
        {
            foreach (string fname in argFnames)
            {
                // データセットにファイルを読み込み
                CM項目DataSet ds = new CM項目DataSet();
                ds.ReadXml(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Model", fname + ".xml"));

                CMCmdSetting cmdSetting = new CMCmdSetting();

                // テーブル名を設定
                cmdSetting.Name = ds.項目一覧[0].項目一覧ID;

                // パラメータ設定
                List<CMCmdParam> paramList = new List<CMCmdParam>();
                foreach (var row in ds.項目)
                {
                    // 更新対象外は無視
                    if (row.入力制限 == "不可") continue;

                    CMCmdParam cmdParam = new CMCmdParam();
                    cmdParam.Name = row.項目名;
                    cmdParam.DbType = (CMDbType)Enum.Parse(typeof(CMDbType), row.項目型);
                    cmdParam.IsKey = row.主キー;
                    cmdParam.IsDecimal = row.小数桁 > 0;

                    paramList.Add(cmdParam);
                }
                cmdSetting.ColumnParams = paramList.ToArray();

                // 設定を追加
                CmdSettings.Add(cmdSetting);
            }
        }
    }
}

File: VB/Lib/CMCommon/DA/CMCmdSettings.vb
Imports System.IO
Imports System.ComponentModel

Imports NEXS.ERP.CM.Common

Namespace DA
    ''' <summary>
    ''' SqlCommand設定コレクションクラス
    ''' </summary>
    Public Class CMCmdSettings
#Region "プロパティ"
        <Category("共通部品")> _
        <Description("SqlCommand設定のコレクション")> _
        Public Property CmdSettings As List(Of CMCmdSetting)
#End Region

#Region "インデクサ"
        ''' <summary>指定IndexのSqlCommand設定を返します。</summary>
        Default Public ReadOnly Property Item(argIndex As Integer) As CMCmdSetting
            Get
                Return CmdSettings(argIndex)
            End Get
        End Property

        ''' <summary>指定のテーブル名のSqlCommand設定を返します。</summary>
        Default Public ReadOnly Property Item(argName As String) As CMCmdSetting
            Get
                Dim result = From row In CmdSettings Where row.Name = argName
                Return result.First()
            End Get
        End Property
#End Region

#Region "コンストラクタ"
        ''' <summary>
        ''' コンストラクタ
        ''' </summary>
        Public Sub New()
            CmdSettings = New List(Of CMCmdSetting)()
        End Sub

        ''' <summary>
        ''' コンストラクタ
        ''' </summary>
        ''' <param name="argFname">読み込むXMLファイル名(拡張子なし)</param>
        Public Sub New(argFname As String)
            Me.New()
            AddFomXml(argFname)
        End Sub
#End Region

        ''' <summary>
        ''' SqlCommand設定をXMLファイルから追加する。
        ''' </summary>
        ''' <param name="argFnames">読み込むXMLファイル名(拡張子なし)</param>
        Public Sub AddFomXml(ParamArray argFnames As String())
            For Each fname As String In argFnames
                ' データセットにファイルを読み込み
                Dim ds As New CMEntityDataSet()
                ds.ReadXml(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Model", fname & ".xml"))

                Dim cmdSetting As New CMCmdSetting()

                ' テーブル名を設定
                cmdSetting.Name = ds.エンティティ(0).テーブル名

                ' パラメータ設定
                Dim paramList As New List(Of CMCmdParam)()
                For Each row As CMEntityDataSet.項目Row In ds.項目
                    ' 更新対象外は無視
                    If row.更新対象外 Then
                        Continue For
                    End If

                    Dim cmdParam As New CMCmdParam()
                    cmdParam.Name = row.項目名
                    cmdParam.DbType = DirectCast([Enum].Parse(GetType(CMDbType), row.項目型), CMDbType)
                    cmdParam.IsKey = row.Key
                    cmdParam.SourceColumn = row.SourceColumn

                    paramList.Add(cmdParam)
                Next
                cmdSetting.ColumnParams = paramList.ToArray()

                ' 設定を追加
                CmdSettings.Add(cmdSetting)
            Next
        End Sub
    End Class
End Namespace
